universal-hash/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+//! Traits for [Universal Hash Functions].
+//!
+//! Universal hash functions select from a "universal family" of possible
+//! hash functions selected by a key. They are well suited to the purpose
+//! of "one time authenticators" for a sequence of bytestring inputs,
+//! as their construction has a number of desirable properties such as
+//! pairwise independence as well as amenability to efficient implementations,
+//! particularly when implemented using SIMD instructions.
+//!
+//! When combined with a cipher, such as in Galois/Counter Mode or the
+//! Salsa20 family AEAD constructions, they can provide the core functionality
+//! for a Message Authentication Code (MAC).
+//!
+//! [Universal Hash Functions]: https://en.wikipedia.org/wiki/Universal_hashing
+
+#![no_std]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+use generic_array::{ArrayLength, GenericArray};
+use generic_array::typenum::Unsigned;
+use subtle::{Choice, ConstantTimeEq};
+
+/// The `UniversalHash` trait defines a generic interface for universal hash
+/// functions.
+pub trait UniversalHash: Clone {
+    /// Size of a block (e.g. field element) this universal hash operates on
+    type BlockSize: ArrayLength<u8>;
+
+    /// Instantiate a universal hash function with the given key
+    fn new(key: Block<Self::BlockSize>) -> Self;
+
+    /// Input a block into the universal hash function
+    fn update_block(&mut self, block: Block<Self::BlockSize>);
```

Comment:
Isn't `Block` is unnecessary here? We don't need const eq in update block, plus IIUC with the current code implementations will not be able to extract the underlying buffer. Plus I think it may be better to pass a reference (i.e. `&GenericArray<u8, Self::BlockSize>`), so code which puts data into XMM register will be able to write `_mm_loadu_si128(block.as_ptr() as *const __m128i)`. Same for the `new` method.

Reply:
It was originally a more convenient way to represent the `Copy` bound.

I can make both `new(key: ...)` and `update_block(...` take a `GenericArray` instead, and change the `Block` type to be `Output`. Does that sound good to you?

Reply:
Yes, looks good, also don't forget `verify`. What do you think about using references?

Reply:
Updated as discussed